### PR TITLE
Work around for multiple displays with respective scales

### LINKED ISSUE #15: Touch input offset on secondary monitor
Hello,
First of all, thank you for this tool.

I know I don't have an usual setup, but here's a bug with it.

This is my setup, I'm on macOS Sierra 10.12.4 and I have two screens:
screens
- The landscape one is the main display for the OS and is 2880x1800
- The portrait one is a secondary display and is 1080x1920 (This is where my IDE lives)

Now, on the main display, everything works fine with scrcpy, but when the window is on the secondary display, there is an offset for the touch input:
gif

### PR DESCRIPTION
**Pull request by wsxyeah**
It's just a workaround to resolve #15

What this pr did is to re-set a new logical size when the window scale factor got changed.
In consequence, the subsequent SDL events will get the coordinates what we wants,
which are consistent across all the displays even if there are different scales.

### CODE REVIEW

**Comment by rom1v** [comment]
(just to confirm I saw the PR, I will try to review soon, thank you)

**Comment by rom1v** [comment]
Ping @AoDevBlue @stek29 #57

File: app/src/scrcpy.c [hunk 1]
```diff
@@ -84,6 +84,17 @@ static SDL_bool event_loop(void) {
...
+                        get_window_scale(SDL_GetWindowFromID(event.button.windowID), &scale_x, &scale_y);
+                        if (scale_x != screen.scale_x || scale_y != screen.scale_y) {
+                            int logical_width, logical_height;
+                            SDL_RenderGetLogicalSize(screen.renderer, &logical_width, &logical_height);
+                            logical_width *= (float)scale_x / screen.scale_x;
+                            logical_height *= (float)scale_y / screen.scale_y;
+                            screen.scale_x = scale_x;
+                            screen.scale_y = scale_y;
+                            SDL_RenderSetLogicalSize(screen.renderer, logical_width, logical_height);
+                        }
```
[the remark this answers is not in the document]
Comment (rom1v) [reply app/src/scrcpy.c]: Oh, but I'm stupid, you need it there so that it changes when you moves the window from/to the secondary screen (it still needs be done in `screen.c` though).

**Comment by AoDevBlue** [comment]
I just checked on my setup and it works on both screens without the workaround :+1:

**Comment by rom1v** [comment]
@AoDevBlue Thank you for your feedbacks. And does the workaround break it?

**Comment by AoDevBlue** [comment]
Yes, this branch works on both screens (vertical without hdpi and horizontal with hdpi) with both the default build config

```
meson x --buildtype release --strip -Db_lto=true
```

and the old workaround one

```
meson x --buildtype release -Dhidpi_support=false
```

File: app/src/screen.h [hunk 2]
```diff
@@ -66,4 +68,6 @@ void screen_resize_to_fit(struct screen *screen);
 // resize window to 1:1 (pixel-perfect)
 void screen_resize_to_pixel_perfect(struct screen *screen);
 
+void get_window_scale(SDL_Window *window, int *scale_x, int *scale_y);
```
[the remark this answers is not in the document]
Comment (wsxyeah) [reply app/src/screen.h]: I don't get any non-integral scale display device for now 😅.
So I will just turn the variables into float, but some tests on such devices are still necessary.

[the remark this answers is not in the document]
Comment (rom1v) [reply app/src/screen.h]: There is no need to use floating point values here (the ratio will lose precision), you can use a "rational" instead (cf [here](https://github.com/Genymobile/scrcpy/commit/ae87885ffafbf140631096a8f6c76fc5df22cfb1#diff-f2a17bd4a20de01474e6482ef776d056R7])

**Comment by stek29** [comment]
Do you want me to test this on macOS?
Do I need to pass any flags to meson? Can I have a tl;dr please? :)

File: app/src/screen.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (wsxyeah) [reply app/src/screen.h]: Do you mean use `rational` to denote the scale factor?
But we still have to convert it to float when performing the new logical size scale calculation.

**Comment by rom1v** [comment]
@ jonalmeida Thank you very much :+1:

> I've found @wsxyeah 's patch to work perfectly for me

Great, good news :+1:

> but unfortunately all touch events fail for your patch

In my patch, I did not updated the logical size on screen change (https://github.com/Genymobile/scrcpy/pull/245#discussion_r215034843). Maybe that if you add this additional patch, it may work (if there are no other mistakes):

```diff
diff --git a/app/src/scrcpy.c b/app/src/scrcpy.c
index 54a7f99..2e2e911 100644
--- a/app/src/scrcpy.c
+++ b/app/src/scrcpy.c
@@ -88,6 +88,7 @@ static SDL_bool event_loop(void) {
                 switch (event.window.event) {
                     case SDL_WINDOWEVENT_EXPOSED:
                     case SDL_WINDOWEVENT_SIZE_CHANGED:
+                        screen_update_scale(&screen);
                         screen_render(&screen);
                         break;
                 }
diff --git a/app/src/screen.c b/app/src/screen.c
index 1ae6c36..1666a78 100644
--- a/app/src/screen.c
+++ b/app/src/screen.c
@@ -313,3 +313,7 @@ void screen_resize_to_pixel_perfect(struct screen *screen) {
         LOGD("Resized to pixel-perfect");
     }
 }
+
+void screen_update_scale(struct screen *screen) {
+    render_set_scaled_logical_size(screen, screen->frame_size);
+}
diff --git a/app/src/screen.h b/app/src/screen.h
index 13103ea..67dcb29 100644
--- a/app/src/screen.h
+++ b/app/src/screen.h
@@ -66,4 +66,8 @@ void screen_resize_to_fit(struct screen *screen);
 // resize window to 1:1 (pixel-perfect)
 void screen_resize_to_pixel_perfect(struct screen *screen);
 
+// recompute the scale in case the window moved from/to another screen with a
+// different HiDPI
+void screen_update_scale(struct screen *screen);
+
 #endif
```

(This is frustrating not to be able to reproduce/test.)

> There were a few build warnings but it seemed to have built properly with the default build instructions.

Thanks, fixed. https://github.com/Genymobile/scrcpy/commit/eca99d5af7d27115bb232f01826d3c5fc4fedd47

> Split-View also seems to work, but there is some bugginess with re-sizing the view: clicking the split partition immediately takes you out of split-view instead of allowing you to alter it.

I'm not sure to understand how the render size have an effect on the split-view behavior.

> Without this patch, it works fine (event though the touch offsets is still a problem).

Arf, this is weird. :disappointed:

File: app/src/scrcpy.c [hunk 3]
```diff
@@ -83,6 +83,9 @@ static SDL_bool event_loop(void) {
                 switch (event.window.event) {
                     case SDL_WINDOWEVENT_EXPOSED:
                     case SDL_WINDOWEVENT_SIZE_CHANGED:
+                        if (!screen_update_logical_size_if_needed(&screen)) {
```
[the remark this answers is not in the document]
Comment (rom1v) [reply app/src/scrcpy.c]: IIRC, for rendering at least, `SDL_WINDOWEVENT_EXPOSED` was necessary when switching fullscreen on some platforms. I'm not sure it's necessary for detecting scaling changes.

File: app/src/screen.c [hunk 4]
```diff
@@ -65,6 +65,33 @@ static void set_window_size(struct screen *screen, struct size new_size) {
...
+    int output_w, output_h;
+    SDL_GL_GetDrawableSize(window, &output_w, &output_h);
+    *scale_x = (float)output_w / win_w;
+    *scale_y = (float)output_h / win_h;
+}
+
+SDL_bool screen_update_logical_size_if_needed(struct screen *screen) {
+    float scale_x, scale_y;
+    get_window_scale(screen->window, &scale_x, &scale_y);
+    if (scale_x != screen->scale_x || scale_y != screen->scale_y) {
```
[the remark this answers is not in the document]
Comment (wsxyeah) [reply app/src/screen.c]: I agree that comparing float values for equality is fragile.

Updating logical size when the frame size changes is ok, but updating on `SDL_WINDOWEVENT_SIZE_CHANGED` is unnecessary.

[the remark this answers is not in the document]
Comment (rom1v) [reply app/src/screen.c]: OK, then we can keep the last width and height passed to `SDL_RenderSetLogicalSize()` (instead of the scale), and call it only when the parameters have changed.

**Comment by jonalmeida** [comment]
@rom1v I'll apply this last patch on your branch and give you an update on Monday.

Regarding the split-view, I'm not sure if that's something that can be solved easily without a reproducing environment for yourself. If I get a chance, I'll try to get some repro steps for you in a separate issue. Maybe someone else will have a better chance of fixing it then. 👍

**Comment by jonalmeida** [comment]
@rom1v I've tried your branch `hidpiscale` along with the patch you suggested, but it unfortunately still has issues. Touch events work on the macbook's screen, but when I move it to the external (low-density) monitor, they stop working.

Also, the release build was failing for lint warnings so I tested it using the debug variant instead:
```
➜  x git:(ae87885) ✗ ninja
[23/31] Compiling C object 'app/app@@scrcpy@exe/src_main.c.o'.
../app/src/main.c:255:5: warning: 'av_register_all' is deprecated [-Wdeprecated-declarations]
    av_register_all();
    ^
/usr/local/Cellar/ffmpeg/4.0.2/include/libavformat/avformat.h:2024:1: note: 'av_register_all' has been explicitly marked deprecated here
attribute_deprecated
^
/usr/local/Cellar/ffmpeg/4.0.2/include/libavutil/attributes.h:94:49: note: expanded from macro 'attribute_deprecated'
#    define attribute_deprecated __attribute__((deprecated))
                                                ^
1 warning generated.
[31/31] Generating scrcpy-server with a custom command.
FAILED: server/scrcpy-server.jar
/Users/jalmeida/src/scrcpy/server/./scripts/build-wrapper.sh ../server/. server/scrcpy-server.jar release
:server:preBuild UP-TO-DATE
:server:preReleaseBuild UP-TO-DATE
:server:compileReleaseAidl UP-TO-DATE
:server:compileReleaseRenderscript UP-TO-DATE
:server:checkReleaseManifest UP-TO-DATE
:server:generateReleaseBuildConfig UP-TO-DATE
:server:prepareLintJar UP-TO-DATE
:server:generateReleaseResValues UP-TO-DATE
:server:generateReleaseResources UP-TO-DATE
:server:mergeReleaseResources UP-TO-DATE
:server:createReleaseCompatibleScreenManifests UP-TO-DATE
:server:processReleaseManifest UP-TO-DATE
:server:splitsDiscoveryTaskRelease UP-TO-DATE
:server:processReleaseResources UP-TO-DATE
:server:generateReleaseSources UP-TO-DATE
:server:javaPreCompileRelease UP-TO-DATE
:server:compileReleaseJavaWithJavac UP-TO-DATE
:server:compileReleaseNdk NO-SOURCE
:server:compileReleaseSources UP-TO-DATE
:server:lintVitalRelease FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':server:lintVitalRelease'.
> java.lang.IllegalStateException: Expected a name but was STRING at line 1 column 99 path $[0].apkInfo.versionName

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
15 actionable tasks: 1 executed, 14 up-to-date
ninja: build stopped: subcommand failed.
```

I also tested out your other branch `logicalscale` but for that, touch events weren't working for both screens. 😞 

Please ping me if there's anything else you'd like me to test out - I'd be happy to help!

**Comment by rom1v** [comment]
> Touch events work on the macbook's screen, but when I move it to the external (low-density) monitor, they stop working.

OK, thank you for your feedbacks.

Instead of "stop working", I guess their coordinates are scaled. For instance, if you click at (100,200), it will generate a click at (200, 400) (or the reverse). (0,0) is at top-left. See gif in https://github.com/Genymobile/scrcpy/issues/15.

> ```
> ../app/src/main.c:255:5: warning: 'av_register_all' is deprecated [-Wdeprecated-declarations]
>     av_register_all();
>     ^
> /usr/local/Cellar/ffmpeg/4.0.2/include/libavformat/avformat.h:2024:1: note: 'av_register_all' has been explicitly marked deprecated here
> attribute_deprecated
> ^
> /usr/local/Cellar/ffmpeg/4.0.2/include/libavutil/attributes.h:94:49: note: expanded from macro 'attribute_deprecated'
> #    define attribute_deprecated __attribute__((deprecated))
>                                                ^
> 1 warning generated.
> ```

This warning is supposed to be fixed by fca806e0954b43cfb2a6b21e66cc9286b713b6a2. I'm interested in your value of `LIBAVCODEC_VERSION_INT`.

> ```
> * What went wrong:
> Execution failed for task ':server:lintVitalRelease'.
> > java.lang.IllegalStateException: Expected a name but was STRING at line 1 column 99 path $[0].apkInfo.versionName
> ```

I cannot reproduce :(

> I also tested out your other branch logicalscale but for that, touch events weren't working for both screens.

That's weird, because `logicalscale` is `dev` + my 2 patches you tested above. The working events probably depend on other factors then (the screen where you start, whether you resize the window or not, etc.).

> Please ping me if there's anything else you'd like me to test out - I'd be happy to help!

Thank you :wink:

**Comment by jonalmeida** [comment]
> Instead of "stop working", I guess their coordinates are scaled. For instance, if you click at (100,200), it will generate a click at (200, 400) (or the reverse). (0,0) is at top-left. See gif in #15.

Sorry about not being clear. By stopped working, I meant the scaling issue still existed.

> Also, when you say "touch events", you mean "click events", right? Touch events may not work as expected (#22).

Correct, click events.


> The working events probably depend on other factors then (the screen where you start, whether you resize the window or not, etc.).

You're right! I closed my laptop screen, and then opened scrcpy on the external monitor where it works as expected.

Re-opening the laptop screen confirms the actual clickable surface area as you can see in the screenshot below:

screen shot 2018-09-18 at 2 27 02 pm

**Comment by rom1v** [comment]
Interesting.

> You're right! I closed my laptop screen, and then opened scrcpy on the external monitor where it works as expected.

Which branch?

**Comment by jonalmeida** [comment]
> Which branch?

`origin/hidpiscale` with both the two patches from the PR comments applied. 

Ideally, it would be nice to have a branch that has everything applied so that I can verify that I've got the right setup.

**Comment by rom1v** [comment]
> `origin/hidpiscale` with both the two patches from the PR comments applied.

OK, and this PR, it works in every cases?

**Comment by jonalmeida** [comment]
> OK, and this PR, it works in every cases?

Correct.

**Comment by halxinate** [comment]
I'd be happy to help with testing, but have no means of building on my side. Is there a way to publish betas for brew?

**Comment by rom1v** [comment]
Does SDL 2.0.9 fix the issue on MacOS?

There is a commit from Oct 12th: [`metal: Fix high dpi and resizing on macOS, and clean up iOS code. Fixes bug #4250.`](https://hg.libsdl.org/SDL/rev/758d4e1222a7?revcount=460).

**Comment by Qormix** [comment]
I don't believe it does. 
I just built and ran master with SDL 2.0.9 installed and the issue is still occurring.

**Comment by fleytman** [comment]
Really looking forward to the fix. When I test the application, I feel uncomfortable that scrcpy occupies the working space of the laptop, and not the second monitor.

**Comment by rom1v** [comment]
Hi,

This PR has been left aside (partly my fault) and has conflicts with current master.

A new PR attempts to fix the issue (#848). But I plan to avoid using the "logical size" at all, to compute correct mouse coordinates directly (if it works). For now, I have a branch, which works without hidpi, but still have problems with hidpi on secondary monitor: see https://github.com/Genymobile/scrcpy/pull/848#issuecomment-544282166.

@wsxyeah @jonalmeida your help is welcome :)

Thank you

**Comment by rom1v** [comment]
Please check #1408, it should fix the issue.
